python/python/bystro/ancestry/ancestry_types.py

```diff
@@ -0,0 +1,158 @@
+"""Classes for common shapes of data in ancestry."""
+
+import attrs
+from attrs import field
+from attrs.validators import ge, instance_of, le
+
+# Attr classes can throw either TypeError or ValueError upon receipt
+# of bad data.  Generally we won't care which, so define the union of
+# the two exceptions to catch all validation errors.  Because we'll be
+# using it in try blocks, this wants to be a tuple rather than a union.
+
+AttrValidationError = (ValueError, TypeError)
+
+
+#  It seems a bit overkill to have an attr class just for vcfs, but
+#  the validator needs to be shared between classes. So it just lives
+#  here as a pure python function with a weird type signature in order
+#  to support what attr expects from its validators.
+
+
+def _vcf_validator(_self: object, _attribute: attrs.Attribute, value: str) -> None:
+    if not isinstance(value, str):
+        err_msg = f"vcf_path must be of type str, got: {value} instead"
+        raise TypeError(err_msg)
+    if "vcf" not in value:
```

Comment:
@akotlar loosened up the validator according to my interpretation of the intent of the regex, lmk if this should rather be a list of predefined extensions?

(fyi testing for string inclusion is ~5x faster than the compiled regex in this case :))

Reply:
hah, man regex is slower than I thought

Reply:
🐢